includes/migration-guide/retargeting/winforms/accessibility-improvements-windows-forms-controls.md

```diff
@@ -3,7 +3,7 @@
olStripMenuItem> controls with their <xref:System.Windows.Forms.ToolStripMenuItem.Checked> property set to <strong>true</strong> now render the associated check mark in a contrasting system color.  Previously the check mark color was not contrasting enough and not visible in High Contrast themes.</li></ul>NOTE: Windows 10 has changed values for some high contrast system colors. Windows Forms Framework is based on the Win32 framework. For the best experience, run on the latest version of Windows and opt in to the latest OS changes by adding an app.manifest file in a test application and uncommenting the following code:<pre><code class="lang-xml">&lt;!-- Windows 10 --&gt;&#13;&#10;&lt;supportedOS Id=&quot;{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}&quot; /&gt;&#13;&#10;</code></pre></br><strong>Improved keyboard navigation</strong><ul><li>When a <xref:System.Windows.Forms.ComboBox> control has its <xref:System.Windows.Forms.ComboBox.DropDownStyle> property set to <xref:System.Windows.Forms.ComboBoxStyle.DropDownList?displayProperty=nameWithType> and is the first control in the tab order on the form, it now displays a focus rectangle when the parent form is opened using the keyboard. Before this change, keyboard focus was on this control, but a focus indicator was not rendered.</li></ul><strong>Improved Narrator support</strong><ul><li>The <xref:System.Windows.Forms.MonthCalendar> control has added support for assistive technologies to access the control, including the ability for Narrator to read the value of the control when previously it could not.</li><li>The <xref:System.Windows.Forms.CheckedListBox> control now notifies Narrator when a <xref:System.Windows.Forms.CheckBox.CheckState?displayProperty=nameWithType> property has been changed. Previously, Narrator did not receive notification and as a result users would not be informed that the <xref:System.Windows.Forms.CheckBox.CheckState> property had been updated.</li><li>The <xref:System.Windows.Forms.LinkLabel> control has changed the way it notifies Narrator of the text of in the control. Previously, Narrator announced this text twice and read &quot;&amp;&quot; symbols as real text even though they are not visible to a user. The duplicated text was removed from the Narrator announcements, as well as unnecessary &quot;&amp;&quot; symbols.</li><li>The <xref:System.Windows.Forms.DataGridViewCell> control types now correctly report the read-only status to Narrator and other assistive technologies.</li><li>Narrator is now able to read the System Menu of child windows in [Multiple-Document Interface]~/docs/framework/winforms/advanced/multiple-document-interface-mdi-applications.md) applications.</li><li>Narrator is now able to read <xref:System.Windows.Forms.ToolStripMenuItem> controls with a <xref:System.Windows.Forms.ToolStripItem.Enabled?displayProperty=nameWithType> property set to <strong>false</strong>. Previously, Narrator was unable to focus on disabled menu items to read the content.</li></ul>|
```

Comment:
```suggestion
|Suggestion|<strong>How to opt in or out of these changes</strong></br>In order for the application to benefit from these changes, it must run on the .NET Framework 4.7.1 or later. The application can benefit from these changes in either of the following ways:<ul><li>It is recompiled to target the .NET Framework 4.7.1. These accessibility changes are enabled by default on Windows Forms applications that target the .NET Framework 4.7.1 or later.</li><li>It opts out of the legacy accessibility behaviors by adding the following [AppContext switch](~/docs/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element.md) to the <code>&lt;runtime&gt;</code> section of the app.config file and setting it to <code>false</code>, as the following example shows.</li></ul><pre><code class="lang-xml">&lt;?xml version=&quot;1.0&quot; encoding=&quot;utf-8&quot;?&gt;&#13;&#10;&lt;configuration&gt;&#13;&#10;&lt;startup&gt;&#13;&#10;&lt;supportedRuntime version=&quot;v4.0&quot; sku=&quot;.NETFramework,Version=v4.7&quot;/&gt;&#13;&#10;&lt;/startup&gt;&#13;&#10;&lt;runtime&gt;&#13;&#10;&lt;!-- AppContextSwitchOverrides value attribute is in the form of &#39;key1=true/false;key2=true/false  --&gt;&#13;&#10;&lt;AppContextSwitchOverrides value=&quot;Switch.UseLegacyAccessibilityFeatures=false&quot; /&gt;&#13;&#10;&lt;/runtime&gt;&#13;&#10;&lt;/configuration&gt;&#13;&#10;</code></pre></br>Applications that target the .NET Framework 4.7.1 or later and want to preserve the legacy accessibility behavior can opt in to the use of legacy accessibility features by explicitly setting this AppContext switch to <code>true</code>.</p>For an overview of UI automation, see the [UI Automation Overview](~/docs/framework/ui-automation/ui-automation-overview.md).&lt;/p/&gt;</p><strong>Added support for UI Automation patterns and properties</strong><br/>Accessibility clients can take advantage of new WinForms accessibility functionality by using common, publicly described invocation patterns. These patterns are not WinForms-specific. For instance, accessibility clients can call the QueryInterface method on the IAccessible interface (MAAS) to obtain an IServiceProvider interface. If this interface is available, clients can use its QueryService method to request an IAccessibleEx interface. For more information, see [Using IAccessibleEx from a Client](https://docs.microsoft.com/windows/desktop/WinAuto/using-iaccessibleex-from-a-client). Starting with the .NET Framework 4.7.1, IServiceProvider and [IAccessibleEx](https://docs.microsoft.com/windows/desktop/WinAuto/iaccessibleex) (where applicable) are available for WinForms accessibility objects.</p>The .NET Framework 4.7.1 adds support for the following UI automation patterns and properties:<ul><li>The <xref:System.Windows.Forms.ToolStripSplitButton> and <xref:System.Windows.Forms.ComboBox> controls support the [Expand/Collapse pattern](~/docs/framework/ui-automation/implementing-the-ui-automation-expandcollapse-control-pattern.md).</li><li>The <xref:System.Windows.Forms.ToolStripMenuItem> control has a [ControlType](~/docs/framework/ui-automation/ui-automation-support-for-the-menubar-control-type.md) property value <xref:System.Windows.Automation.ControlType.MenuItem?displayProperty=nameWithType>.</li><li>The <xref:System.Windows.Forms.ToolStripItem> control supports the <xref:System.Windows.Automation.AutomationElement.NameProperty> property and the [Expand/Collapse pattern](~/docs/framework/ui-automation/implementing-the-ui-automation-expandcollapse-control-pattern.md).</li><li>The <xref:System.Windows.Forms.ToolStripDropDownItem> control supports <xref:System.Windows.Forms.AccessibleEvents> indicating StateChange and NameChange when drop down is expanded or collapsed.</li><li>The <xref:System.Windows.Forms.ToolStripDropDownButton> control has a [ControlType](~/docs/framework/ui-automation/ui-automation-support-for-the-menubar-control-type.md) property value of <xref:System.Windows.Automation.ControlType.MenuItem?displayProperty=nameWithType>.</li><li>The <xref:System.Windows.Forms.DataGridViewCheckBoxCell> control supports the <xref:System.Windows.Automation.TogglePattern>.</li><li>The <xref:System.Windows.Forms.NumericUpDown> and <xref:System.Windows.Forms.DomainUpDown> controls support the <xref:System.Windows.Automation.AutomationElement.NameProperty> property and have a [ControlType](~/docs/framework/ui-automation/ui-automation-support-for-the-spinner-control-type.md) of <xref:System.Windows.Automation.ControlType.Spinner?displayProperty=nameWithType>.</p></li></ul><strong>Improvements to the PropertyGrid control</strong></br>The .NET Framework 4.7.1 adds the following improvements to the PropertyBrowser control:<ul><li>The <strong>Details</strong> button in the error dialog that is displayed when the user enters an incorrect value in the <xref:System.Windows.Forms.PropertyGrid> control supports the [Expand/Collapse pattern](~/docs/framework/ui-automation/implementing-the-ui-automation-expandcollapse-control-pattern.md), state and name change notifications, and a [ControlType]~/docs/framework/ui-automation/ui-automation-support-for-the-menubar-control-type.md) property with a value of <xref:System.Windows.Automation.ControlType.MenuItem?displayProperty=nameWithType>.</li><li>The message pane displayed when the <strong>Details</strong> button of the error dialog is expanded is now keyboard accessible and allows Narrator to announce the content of the error message.</li><li>The <xref:System.Windows.Forms.AccessibleRole> of rows in the <xref:System.Windows.Forms.PropertyGrid> control have changed from &quot;Row&quot; to &quot;Cell&quot;. The cell maps to UIA ControlType &quot;DataItem&quot;, which allows it to support appropriate keyboard shortcuts and Narrator announcements.</li><li>The <xref:System.Windows.Forms.PropertyGrid> control rows that represent header items when the <xref:System.Windows.Forms.PropertyGrid> control has a <xref:System.Windows.Forms.PropertyGrid.PropertySort> property set to <xref:System.Windows.Forms.PropertySort.Categorized?displayProperty=nameWithType> have a [ControlType](~/docs/framework/ui-automation/ui-automation-support-for-the-menubar-control-type.md) property value of <xref:System.Windows.Automation.ControlType.Button?displayProperty=nameWithType>.</li><li>The <xref:System.Windows.Forms.PropertyGrid> control rows that represent header items when the <xref:System.Windows.Forms.PropertyGrid> control has a <xref:System.Windows.Forms.PropertyGrid.PropertySort>  property set to <xref:System.Windows.Forms.PropertySort.Categorized?displayProperty=nameWithType> support the [Expand/Collapse pattern](~/docs/framework/ui-automation/implementing-the-ui-automation-expandcollapse-control-pattern.md).</li><li>Improved keyboard navigation between the grid and the ToolBar above it. Pressing &quot;Shift-Tab&quot; now selects the first ToolBar button, instead of the whole ToolBar.</li><li><xref:System.Windows.Forms.PropertyGrid> controls displayed in High Contrast mode will now draw a focus rectangle around the ToolBar button which corresponds to the current <xref:System.Windows.Forms.PropertyGrid.PropertySort> property value.</li><li><xref:System.Windows.Forms.PropertyGrid> controls displayed in High Contrast mode and with a <xref:System.Windows.Forms.PropertyGrid.PropertySort> property set to <xref:System.Windows.Forms.PropertySort.Categorized?displayProperty=nameWithType> will now display the background of category headers in a highly contrasting color.</li><li><xref:System.Windows.Forms.PropertyGrid> controls better differentiates between ToolBar items with focus and the ToolBar items which indicate the current value of the <xref:System.Windows.Forms.PropertyGrid.PropertySort> property. This fix consists of a High Contrast change and a change for non-High Contrast scenarios.</li><li><xref:System.Windows.Forms.PropertyGrid> control ToolBar items which indicates the current value of the <xref:System.Windows.Forms.PropertyGrid.PropertySort> property support the <xref:System.Windows.Automation.TogglePattern>.</li><li>Improved Narrator support for distinguishing the selected alignment in the Alignment Picker.</li><li>When an empty <xref:System.Windows.Forms.PropertyGrid> control is displayed on a form, it will now receive focus where previously it would not.</p></li></ul><strong>Use of OS-defined colors in High Contrast themes</strong></br><ul><li>The <xref:System.Windows.Forms.Button> and <xref:System.Windows.Forms.CheckBox> controls with their <xref:System.Windows.Forms.ButtonBase.FlatStyle> property set to <xref:System.Windows.Forms.FlatStyle.System?displayProperty=nameWithType>, which is the default style, now use OS-defined colors in High Contrast theme when selected. Previously, text and background colors were not contrasting and were hard to read.</li><li>The <xref:System.Windows.Forms.Button>, <xref:System.Windows.Forms.CheckBox>, <xref:System.Windows.Forms.RadioButton>, <xref:System.Windows.Forms.Label>, <xref:System.Windows.Forms.LinkLabel> and <xref:System.Windows.Forms.GroupBox> controls with their <xref:System.Windows.Forms.Control.Enabled> property set to <strong>false</strong> used a shaded color to render text in High Contrast themes, resulting in low contrast against the background. Now these controls use the &quot;Disabled Text&quot; color defined by the OS. This fix applies to controls with their <code>FlatStyle</code> property set to a value other than <xref:System.Windows.Forms.FlatStyle.System?displayProperty=nameWithType>. The latter controls are rendered by the OS.</li><li><xref:System.Windows.Forms.DataGridView> now renders a visible rectangle around the content of the cell which has the current focus. Previously, this was not visible in certain High Contrast themes.</li><li><xref:System.Windows.Forms.ToolStripMenuItem> controls with their <xref:System.Windows.Forms.ToolStripMenuItem.Enabled> property set to <strong>false</strong> now use the &quot;Disabled Text&quot; color defined by the OS.</li><li><xref:System.Windows.Forms.ToolStripMenuItem> controls with their <xref:System.Windows.Forms.ToolStripMenuItem.Checked> property set to <strong>true</strong> now render the associated check mark in a contrasting system color.  Previously the check mark color was not contrasting enough and not visible in High Contrast themes.</li></ul>NOTE: Windows 10 has changed values for some high contrast system colors. Windows Forms Framework is based on the Win32 framework. For the best experience, run on the latest version of Windows and opt in to the latest OS changes by adding an app.manifest file in a test application and uncommenting the following code:<pre><code class="lang-xml">&lt;!-- Windows 10 --&gt;&#13;&#10;&lt;supportedOS Id=&quot;{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}&quot; /&gt;&#13;&#10;</code></pre></br><strong>Improved keyboard navigation</strong><ul><li>When a <xref:System.Windows.Forms.ComboBox> control has its <xref:System.Windows.Forms.ComboBox.DropDownStyle> property set to <xref:System.Windows.Forms.ComboBoxStyle.DropDownList?displayProperty=nameWithType> and is the first control in the tab order on the form, it now displays a focus rectangle when the parent form is opened using the keyboard. Before this change, keyboard focus was on this control, but a focus indicator was not rendered.</li></ul><strong>Improved Narrator support</strong><ul><li>The <xref:System.Windows.Forms.MonthCalendar> control has added support for assistive technologies to access the control, including the ability for Narrator to read the value of the control when previously it could not.</li><li>The <xref:System.Windows.Forms.CheckedListBox> control now notifies Narrator when a <xref:System.Windows.Forms.CheckBox.CheckState?displayProperty=nameWithType> property has been changed. Previously, Narrator did not receive notification and as a result users would not be informed that the <xref:System.Windows.Forms.CheckBox.CheckState> property had been updated.</li><li>The <xref:System.Windows.Forms.LinkLabel> control has changed the way it notifies Narrator of the text of in the control. Previously, Narrator announced this text twice and read &quot;&amp;&quot; symbols as real text even though they are not visible to a user. The duplicated text was removed from the Narrator announcements, as well as unnecessary &quot;&amp;&quot; symbols.</li><li>The <xref:System.Windows.Forms.DataGridViewCell> control types now correctly report the read-only status to Narrator and other assistive technologies.</li><li>Narrator is now able to read the System Menu of child windows in [Multiple-Document Interface]~/docs/framework/winforms/advanced/multiple-document-interface-mdi-applications.md) applications.</li><li>Narrator is now able to read <xref:System.Windows.Forms.ToolStripMenuItem> controls with a <xref:System.Windows.Forms.ToolStripItem.Enabled?displayProperty=nameWithType> property set to <strong>false</strong>. Previously, Narrator was unable to focus on disabled menu items to read the content.</li></ul>|
```